Alex and Happy Part 2/article.md

```diff
@@ -0,0 +1,1372 @@
+
+As such, these libraries have caveats like not being able to handle left-recursion. Their performance can also exponentially degrade due to backtracking.
+
+While comparing the implementations of both parsing techniques is beyond the scope of this post, you can read two suggested blog posts for more information: [From recursive descent to LR parsing](https://www.abubalay.com/blog/2021/12/31/lr-control-flow) and [Which Parsing Approach?](https://tratt.net/laurie/essays/entries/which_parsing_approach.html).
+
+Another difference between these libraries is that Happy is a parser generator.
+It means that instead of writing Haskell files (`.hs`) like you do with Megaparsec, you will write Happy (`.y`) files. Those will generate Haskell files with implementations of the parser recognizing the described grammar.
+
+To sum it up, here are some pros and cons of Happy in comparison to Megaparsec:
+
+Pros:
+* Parser generators report ambiguous grammars to the user, creating greater trust in the parser's output and saving from debugging headaches.
+* Bottom-up parsers can handle left-recursion, while top-down parsers can't.
+* The performance will be more predictable, as Megaparsec can cause exponential time complexity with the use of `try`. In addition, Happy's "static" generated LALR algorithm has better known complexity constraints.
+* Dealing with operator precedence and associativity requires less labor. In Megaparsec, there are standard tools like [`makeExprParser`](https://hackage.haskell.org/package/parser-combinators-1.3.0/docs/Control-Monad-Combinators-Expr.html#v:makeExprParser) to deal with them, but once you need something that is non-standard, you'll need to write more non-trivial parser rules by hand.
+
+Cons:
+* Less idiomatic, as you need to write Happy grammars instead of Haskell code.
+* Less flexible since you can't change Happy’s algorithm. Meanwhile, parser combinators enable you to easily create your own functions.
+
+## How it works
+
+An in-depth explanation of how an LR(1) parser works is well beyond the scope of this tutorial. For this, we recommend you read our article on [how to Implement an LR(1) Parser](https://serokell.io/blog/how-to-implement-lr1-parser).
+We will, however, introduce a few basic concepts required to understand this article better.
+
+### Terminology
+
+A **language** (formally) is a (likely infinite) set of all grammatically correct sentences.
+
+A **grammar** describes how to construct an arbitrary sentence of the language with simple string substitutions.
+Formally, a grammar consists of terminals, non-terminals, production rules and a designated starting non-terminal.
+
+The process of constructing a sentence using the grammar is called a **derivation**.
+
+Parsing is in some sense a process inverse to derivation, i.e. given the end result (the input string in parsing terms) we recover a derivation that produced it.
+
```

Comment:
```suggestion
Sentences are built from symbols, which can be terminals or non-terminals. 
```

I think this would fit well here instead of being in parenthesis.